File: archive/cdktf-py/Pr7948/lib/stacks/api_stack.py
"""API infrastructure - API Gateway with custom domains"""

from constructs import Construct
from cdktf_cdktf_provider_aws.apigatewayv2_api import Apigatewayv2Api
from cdktf_cdktf_provider_aws.apigatewayv2_stage import Apigatewayv2Stage, Apigatewayv2StageAccessLogSettings
from cdktf_cdktf_provider_aws.apigatewayv2_integration import Apigatewayv2Integration
from cdktf_cdktf_provider_aws.apigatewayv2_route import Apigatewayv2Route
from cdktf_cdktf_provider_aws.lambda_permission import LambdaPermission
from cdktf_cdktf_provider_aws.acm_certificate import AcmCertificate, AcmCertificateDomainValidationOptions
from cdktf_cdktf_provider_aws.acm_certificate_validation import AcmCertificateValidation
from cdktf_cdktf_provider_aws.route53_zone import Route53Zone
from cdktf_cdktf_provider_aws.route53_record import Route53Record
from cdktf_cdktf_provider_aws.cloudwatch_log_group import CloudwatchLogGroup
import json


class ApiStack(Construct):
    """Creates API Gateway with custom domain and ACM certificate"""

    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        region: str,
        environment_suffix: str,
        lambda_function
    ):
        super().__init__(scope, construct_id)

        self.region = region
        self.environment_suffix = environment_suffix

        # CloudWatch Log Group for API Gateway
        log_group = CloudwatchLogGroup(
            self,
            "api-logs",
            name=f"/aws/apigateway/dr-api-{region}-{environment_suffix}",
            retention_in_days=7,
            tags={
                "Name": f"dr-api-logs-{region}-{environment_suffix}"
            }
        )

        # API Gateway HTTP API
        self.api_gateway = Apigatewayv2Api(
            self,
            "api",
            name=f"dr-payment-api-{region}-{environment_suffix}",
            protocol_type="HTTP",
            cors_configuration={
                "allow_origins": ["*"],
                "allow_methods": ["GET", "POST", "PUT", "DELETE"],
                "allow_headers": ["*"],
                "max_age": 300
            },
            tags={
                "Name": f"dr-payment-api-{region}-{environment_suffix}"
            }
        )

        # Lambda integration
        integration = Apigatewayv2Integration(
            self,
            "lambda-integration",
            api_id=self.api_gateway.id,
            integration_type="AWS_PROXY",
            integration_uri=lambda_function.arn,
            integration_method="POST",
            payload_format_version="2.0"
        )

        # Routes
        Apigatewayv2Route(
            self,
            "payment-route",
            api_id=self.api_gateway.id,
            route_key="POST /payment",
            target=f"integrations/{integration.id}"
        )

        Apigatewayv2Route(
            self,
            "health-route",
            api_id=self.api_gateway.id,
            route_key="GET /health",
            target=f"integrations/{integration.id}"
        )

        # API Stage
        self.api_stage = Apigatewayv2Stage(
            self,
            "api-stage",
            api_id=self.api_gateway.id,
            name="prod",
            auto_deploy=True,
            access_log_settings=Apigatewayv2StageAccessLogSettings(
                destination_arn=log_group.arn,
                format=json.dumps({
                    "requestId": "$context.requestId",
                    "ip": "$context.identity.sourceIp",
                    "requestTime": "$context.requestTime",
                    "httpMethod": "$context.httpMethod",
                    "routeKey": "$context.routeKey",
                    "status": "$context.status",
                    "protocol": "$context.protocol",
                    "responseLength": "$context.responseLength"
                })
            ),
            tags={
                "Name": f"dr-api-stage-{region}-{environment_suffix}"
            }
        )

        # Lambda permission for API Gateway
        LambdaPermission(
            self,
            "api-lambda-permission",
            statement_id="AllowAPIGatewayInvoke",
            action="lambda:InvokeFunction",
            function_name=lambda_function.function_name,
            principal="apigateway.amazonaws.com",
            source_arn=f"{self.api_gateway.execution_arn}/*/*"
        )

        # API endpoint
        self.api_endpoint = f"{self.api_gateway.api_endpoint}/{self.api_stage.name}"

File: archive/cdktf-ts/Pr3522/lib/api-stack.ts
import { Construct } from 'constructs';
import { Vpc } from '@cdktf/provider-aws/lib/vpc';
import { Alb } from '@cdktf/provider-aws/lib/alb';
import { Apigatewayv2Api } from '@cdktf/provider-aws/lib/apigatewayv2-api';
import { Apigatewayv2Stage } from '@cdktf/provider-aws/lib/apigatewayv2-stage';
import { Apigatewayv2Route } from '@cdktf/provider-aws/lib/apigatewayv2-route';
import { Apigatewayv2Integration } from '@cdktf/provider-aws/lib/apigatewayv2-integration';
import { IamRole } from '@cdktf/provider-aws/lib/iam-role';
import { IamRolePolicy } from '@cdktf/provider-aws/lib/iam-role-policy';
import { LambdaFunction } from '@cdktf/provider-aws/lib/lambda-function';
import { LambdaPermission } from '@cdktf/provider-aws/lib/lambda-permission';
import { DataArchiveFile } from '@cdktf/provider-archive/lib/data-archive-file';
import * as path from 'path';

interface ApiStackProps {
  vpc: Vpc;
  alb: Alb;
  region: string;
  environmentSuffix: string;
}

export class ApiStack extends Construct {
  public readonly websocketApi: Apigatewayv2Api;

  constructor(scope: Construct, id: string, props: ApiStackProps) {
    super(scope, id);

    const lambdaRole = new IamRole(this, 'lambda-role', {
      name: `portfolio-ws-lambda-role-${props.environmentSuffix}`,
      assumeRolePolicy: JSON.stringify({
        Version: '2012-10-17',
        Statement: [
          {
            Action: 'sts:AssumeRole',
            Effect: 'Allow',
            Principal: {
              Service: 'lambda.amazonaws.com',
            },
          },
        ],
      }),
    });

    new IamRolePolicy(this, 'lambda-policy', {
      name: `portfolio-ws-lambda-policy-${props.environmentSuffix}`,
      role: lambdaRole.id,
      policy: JSON.stringify({
        Version: '2012-10-17',
        Statement: [
          {
            Effect: 'Allow',
            Action: [
              'logs:CreateLogGroup',
              'logs:CreateLogStream',
              'logs:PutLogEvents',
            ],
            Resource: 'arn:aws:logs:*:*:*',
          },
          {
            Effect: 'Allow',
            Action: ['execute-api:ManageConnections', 'execute-api:Invoke'],
            Resource: '*',
          },
        ],
      }),
    });

    // Create Lambda deployment package
    const lambdaAsset = new DataArchiveFile(this, 'websocket-lambda-archive', {
      type: 'zip',
      sourceDir: path.join(__dirname, 'lambda', 'websocket-handler'),
      outputPath: path.join(
        __dirname,
        '../.terraform',
        `websocket-handler-${props.environmentSuffix}.zip`
      ),
    });

    const websocketLambda = new LambdaFunction(this, 'websocket-handler', {
      functionName: `portfolio-ws-handler-${props.environmentSuffix}`,
      runtime: 'nodejs18.x',
      handler: 'index.handler',
      role: lambdaRole.arn,
      filename: lambdaAsset.outputPath,
      sourceCodeHash: lambdaAsset.outputBase64Sha256,
      timeout: 30,
      memorySize: 256,
      environment: {
        variables: {
          ALB_DNS: props.alb.dnsName,
        },
      },
    });

    this.websocketApi = new Apigatewayv2Api(this, 'websocket-api', {
      name: `portfolio-ws-api-${props.environmentSuffix}`,
      protocolType: 'WEBSOCKET',
      routeSelectionExpression: '$request.body.action',
      description: 'WebSocket API for real-time portfolio updates',
    });

    const integration = new Apigatewayv2Integration(
      this,
      'websocket-integration',
      {
        apiId: this.websocketApi.id,
        integrationType: 'AWS_PROXY',
        integrationUri: websocketLambda.invokeArn,
        integrationMethod: 'POST',
        connectionType: 'INTERNET',
      }
    );

    new Apigatewayv2Route(this, 'connect-route', {
      apiId: this.websocketApi.id,
      routeKey: '$connect',
      target: `integrations/${integration.id}`,
    });

    new Apigatewayv2Route(this, 'disconnect-route', {
      apiId: this.websocketApi.id,
      routeKey: '$disconnect',
      target: `integrations/${integration.id}`,
    });

    new Apigatewayv2Route(this, 'default-route', {
      apiId: this.websocketApi.id,
      routeKey: '$default',
      target: `integrations/${integration.id}`,
    });

    new Apigatewayv2Stage(this, 'websocket-stage', {
      apiId: this.websocketApi.id,
      name: 'prod',
      autoDeploy: true,
    });

    new LambdaPermission(this, 'websocket-lambda-permission', {
      statementId: 'AllowAPIGatewayInvoke',
      action: 'lambda:InvokeFunction',
      functionName: websocketLambda.functionName,
      principal: 'apigateway.amazonaws.com',
      sourceArn: `${this.websocketApi.executionArn}/*/*`,
    });
  }
}
